File: src/routes/price-list-request.routes.ts
import { Hono } from "hono";
import type { Env } from "@/types/hono";
import { PriceListRequestService } from "@/services/price-list-request.service";
import { successResponse } from "@/utils/response";
import { authMiddleware } from "@/middleware/auth";
import { z } from "zod";

export const priceListRequestRoutes = new Hono<Env>();

let service: PriceListRequestService | null = null;

function getService(): PriceListRequestService {
  if (!service) {
    service = new PriceListRequestService();
  }
  return service;
}

// Apply auth middleware
priceListRequestRoutes.use("*", authMiddleware);

// POST /api/price-list-requests - Submit a price list request (admin users)
priceListRequestRoutes.post("/", async (c) => {
  const user = c.get("user");

  const formData = await c.req.formData();
  const file = formData.get("file") as File;
  const vendorName = formData.get("vendorName") as string;
  const companyName = formData.get("companyName") as string || "";

  if (!file) {
    return c.json({ success: false, error: { message: "File is required" } }, 400);
  }

  if (!vendorName || vendorName.trim().length < 2) {
    return c.json({ success: false, error: { message: "Vendor name is required (min 2 characters)" } }, 400);
  }

  const buffer = Buffer.from(await file.arrayBuffer());

  const result = await getService().submitRequest(
    user,
    companyName,
    {
      vendorName: vendorName.trim(),
      filename: file.name,
      mimeType: file.type,
      data: buffer,
    }
  );

  return c.json(successResponse(result), 201);
});

// GET /api/price-list-requests - List my company's requests (admin)
priceListRequestRoutes.get("/", async (c) => {
  const user = c.get("user");

  const requests = await getService().listRequestsForCompany(user);
  return c.json(successResponse(requests));
});

File: fisoft/src/main/kotlin/routes/PriceListRequestRoutes.kt
package com.ilyasdemirkiran.routes

import com.ilyasdemirkiran.middleware.authenticateUser
import com.ilyasdemirkiran.middleware.currentCompanyId
import com.ilyasdemirkiran.services.PriceListRequestService
import com.ilyasdemirkiran.utils.successResponse
import io.ktor.http.*
import io.ktor.http.content.*
import io.ktor.server.application.*
import io.ktor.server.request.*
import io.ktor.server.response.*
import io.ktor.server.routing.*
import kotlinx.serialization.Serializable

@Serializable
data class CreatePriceListReq(
    val vendorId: String? = "",
    val vendorName: String,
    val companyName: String? = null,
    val message: String? = null
)

fun Route.priceListRequestRoutes() {
    val service = PriceListRequestService()

    route("/api/price-list-requests") {
        get {
            call.authenticateUser()
            val requests = service.listAllRequests()
            call.respond(HttpStatusCode.OK, successResponse(requests))
        }

        post {
            call.authenticateUser()
            val companyId = call.currentCompanyId
            var vendorName = ""
            var companyName = ""
            var vendorId = ""
            var message: String? = null

            if (call.request.contentType().match(ContentType.Application.Json)) {
                val req = call.receive<CreatePriceListReq>()
                vendorName = req.vendorName
                companyName = req.companyName ?: ""
                vendorId = req.vendorId ?: ""
                message = req.message
            } else {
                val multipart = call.receiveMultipart()
                multipart.forEachPart { part ->
                    if (part is PartData.FormItem) {
                        when (part.name) {
                            "vendorName" -> vendorName = part.value
                            "companyName" -> companyName = part.value
                            "vendorId" -> vendorId = part.value
                            "message" -> message = part.value
                        }
                    }
                    part.dispose()
                }
            }

            if (vendorName.isBlank()) {
                call.respond(HttpStatusCode.BadRequest, mapOf("success" to false, "error" to "Vendor name is required"))
                return@post
            }

            val created = service.createRequest(companyId, companyName, vendorId, vendorName, message)
            call.respond(HttpStatusCode.Created, successResponse(created))
        }
    }
}
